prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+"""This module provides automatic hyperparameter selection using optuna."""
+
+from __future__ import annotations  # noqa FI58
+
+import os
+from typing import Any, Optional
+
+import optuna
+import transformers
+from datasets import Dataset, concatenate_datasets
+from optuna.trial import Trial
+from transformers import PreTrainedModel  # noqa
+from transformers import PreTrainedTokenizer
+
+from prompt2model.model_trainer.generate import GenerationModelTrainer
+from prompt2model.param_selector.base import ParamSelector
+from prompt2model.prompt_parser.base import PromptSpec
+from prompt2model.utils.config import DEFAULT_HYPERPARAMETERS_SPACE
+
+
+class OptunaParamSelector(ParamSelector):
+    """Uses optuna for searching for hyperparameters."""
+
+    def __init__(self, trainer: GenerationModelTrainer, n_trials: int):
+        """Initialize with train/val datasets and a prompt specification.
+
+        Args:
+            trainer (BaseTrainer): trainer object from GenerationModelTrainer
+            n_trials (int): The number of trials for conducting hyperparameter search
+        """
+        self.generation_model_trainer = trainer
+        self.n_trials = n_trials
+
+    def optimize_hyperparameters(
+        self,
+        training_datasets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: Optional[dict[str, Any]] = None,
+    ) -> dict[str, Any]:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_datasets (list[Dataset]): One or more training datasets or
+                subset of the actual dataset for running the trainer.
+            validation_sets (Dataset): A dataset for computing validation metrics.
+            hyperparameter_space (Optional[dict[str, Any]], optional): The set
+                of possible values of hyperparaneters values required for doing
+                optimal hyperparameter search. Defaults to None.
+
+        Returns:
+            Returns a dict which contains the best hyperparameters
```

Comment:
Comments should almost always be complete sentences:

```suggestion
            Returns a dict which contains the best hyperparameters.
```

Reply:
Please address this comment before merging!

Reply:
Done